hoomd/md/Thermostat.h

```diff
@@ -0,0 +1,268 @@
+//
+// Created by girard01 on 11/9/22.
+//
+
+#ifndef HOOMD_THERMOSTAT_H
+#define HOOMD_THERMOSTAT_H
+
+#include <array>
+#include <hoomd/HOOMDMath.h>
+#include <hoomd/ParticleGroup.h>
+#include <hoomd/md/ComputeThermo.h>
+#include <hoomd/Variant.h>
+#include "hoomd/RNGIdentifiers.h"
+#include "hoomd/RandomNumbers.h"
+
+namespace hoomd::md {
+    class Thermostat {
+    public:
+        Thermostat(std::shared_ptr<Variant> T) :
+        m_T(T)
+        {}
+
+        virtual ~Thermostat() {}
+
+        virtual std::array<Scalar, 2> getRescalingFactorsOne(uint64_t timestep, Scalar deltaT) {
+            return {Scalar(1.0), Scalar(1.0)};
+        }
+
+        virtual std::array<Scalar, 2> getRescalingFactorsTwo(uint64_t timestep, Scalar deltaT) {
+            return {Scalar(1.0), Scalar(1.0)};
+        }
+
+        virtual void advanceThermostat(uint64_t timestep, Scalar deltaT, bool aniso) {}
+
+        virtual void thermalizeThermostat() {}
+
+        virtual Scalar getThermostatEnergy(uint64_t timestep) {return Scalar(0.0);}
+
+        std::shared_ptr<Variant> getT(){
+            return m_T;
+        }
+
+        void setT(std::shared_ptr<Variant> T){
+            m_T = T;
+        }
+
+        Scalar getTemperature(uint64_t timestep){
+            return (*m_T)(timestep);
+        }
+
+        void attach(std::shared_ptr<ParticleGroup> group, std::shared_ptr<ComputeThermo> thermo, uint16_t seed){
+            if(is_attached)
+                throw std::runtime_error("Thermostats can only be attached to a single integrator");
+            setSeed(seed);
+            is_attached = true;
+            m_group = group;
+            m_thermo = thermo;
+        }
+
+        void detach(){
+            is_attached = false;
+            m_group = nullptr;
+            m_thermo = nullptr;
+        }
+
+        bool attached() const{
+            return is_attached;
+        }
+
+        virtual bool requiresSeed(){
+            return false;
+        }
+
+        virtual void setSeed(uint16_t seed){}
+
+#ifdef ENABLE_MPI
+        virtual void broadcastThermostat(MPI_Comm){}
```

Comment:
I don't know how it got this inconsistent, but the existing codebase broadcasts the thermostat for all code paths *except* NVTMTKGPU. Checking the code history, these broadcasts were present in the original MTK code implementation 9 years ago.

In theory, we shouldn't need the broadcasts. All inputs to the thermostat equations are identical on all ranks, so each rank will compute identical results. If we find that this is not the case in practice then we should find and fix the cause of that before resorting to an expensive MPI broadcast on every time step.

Let's remove these broadcasts.